### ISSUE TITLE
Plotting directed PAGA on basis diffmap

### ISSUE THREAD

**Issue by Marius1311** [theislab/scvelo#242]
I'm plotting a directed PAGA on basis diffmap using `scv.pl.paga(adata, basis='diffmap', color='clusters_coarse')`. This gives me:
![image](https://user-images.githubusercontent.com/24494650/86599216-998b4280-bf9e-11ea-90f4-26dd6e8a6668.png)

so the dots are misplaced. I checked for basis `umap` and `pca`, in both cases, it works fine. 

#### Versions:
Running scvelo 0.2.2.dev51+ga7de78a (python 3.6.9) on 2020-07-06 14:45.
>

**Comment by Marius1311** [comment]
great, thanks Volker!

**Comment by davisidarta** [comment]
This is unrelated to the issue per se: I came across this after looking for solutions for a similar issue I'm having with PAGA graphs. However, I couldn't help to notice this data does not look like the human epithelial cell lineage at all. Perhaps you're using mouse data? If not, only 2 diffusion components are unable to describe this data. [I suggest a look at Fig.7 of this manuscript.](https://www.nature.com/articles/s41598-020-76488-2#Fig7), where we have characterized the epithelial cell lineage with a meta-analysis of different studies.

**Comment by Marius1311** [comment]
Hi @davisidarta, thanks for your comment, this is in fact mouse data and I suppose this isn't enough cells to capture all the lineage dynamics (~1k cells). This was really only to test PAGA on a diffmap.

Correct me if I'm wrong, but the club -> ciliated and club -> goblet transitions shown in this diffmap are as expected for murine epithelial development, no? At least that's what I got from Weinberger, S. E., Cockrill, B. A. & Mandel, J. Principles of Pulmonary Medicine. Philadelphia. (2008).

**Comment by Marius1311** [comment]
Of course, basal cells are missing in this plot due to undersampling.

**Comment by Marius1311** [comment]
Thanks for pointing me to your paper! I will have to go trough this is a bit more detail but it sounds quite interesting. In fact, for the identification of initial and terminal states, we're recently proposed an alternative method called "CellRank" which makes use of RNA velocity, maybe worthwhile to check out: https://cellrank.readthedocs.io/en/latest/

**Comment by davisidarta** [comment]
Hi @Marius1311, thank you for your consideration, and forgive me If I was intrusive. 

Yes, most of your transitions are as expected, and undersampling would be a fine explanation for the missing basal cells, which are reasonably rare compared to other cell types. I understand diffmaps are used broadly for lineage inference and pseudotime estimation, yet [it has been shown ](https://static-content.springer.com/esm/art%3A10.1038%2Fs41587-019-0068-4/MediaObjects/41587_2019_68_MOESM1_ESM.pdf) (take a look at Suppl. Fig. 2b) that several diffusion components are necessary to account for data with various lineages. Perhaps you would also like to check [dbMAP](https://github.com/davisidarta/dbMAP), in which I accelerated and generalized diffusion maps to approximate the Laplace-Beltrami Operator (LPO) and estimate data intrinsic dimensionality, balance the resulting components across all possible diffusion scales, and perform another round of LPO approximation of this high-dimensional diffusion basis with UMAP (still preparing the second manuscript version, I'm on my own on all of this).

CellRank works incredibly well and I'm in fact also using it for our next study, in which we have RNA velocity information. For this lung and alveolar epithelia study, we had to integrate over 100k cells and were in quite a rush, so we started from digital expression matrices instead of aligning everything from scratch and were unable to compute RNA velocity. It would be awesome, however, to be able to get heatmaps of lineage expression trends without the need for RNA velocity information for cases such as these.

P.S.: If you're doing this for a particular study, I'd be happy to share the integrated alveolar human reference as h5ad.

**Comment by Marius1311** [comment]
Hi @davisidarta, wow, thanks so much for your swift, and very informative reply! I'll definitely check out dpMAPs, I like the fact that it's based on NMSLIB, which seems like the current best option to compute KNN graphs - in fact, I think @VolkerBergen has even been looking into HNSW himself. 

Talking about CellRank, you could easily use that with a KNN graph computed from NMSLIB in whatever non-metric space, we take whatever is saved in `adata.p` as our input to compute transition matrices. Re your question of using CellRank without velocities - this is an excellent question, and it's the reason for our modular design of CellRank, structuring it into `kernels` and `estimators`. Basically, `kernels` build transition matrices and `estimators` do fun stuff with them. Our most widely used kernel is the `velocity_kernel`, however, we also have a `palantir_kernel` which mimicks [palantir](https://github.com/dpeerlab/Palantir). For that, you don't need velocity information, just any pseudotime will do! Further `kernels` are likely to come (think of e.g. including multi-model information in transition matrix estimation), and you can even build your own kernel, as has recently been done [here](https://github.com/zsteve/statOT), allowing static optimal transport to be used from within CellRank!

Oh, and thanks for offering to share that data set! However, that's the one where you don't have velocity information, right?

**Comment by davisidarta** [comment]
_Hi davisidarta, wow, thanks so much for your swift, and very informative reply!_

My pleasure!

 _I'll definitely check out dpMAPs, I like the fact that it's based on NMSLIB, which seems like the current best option to compute KNN graphs - in fact, I think @VolkerBergen has even been looking into HNSW himself_

Thank you for your interest! I've implemented a friendly wrapper for it in[ dbmap.ann](https://github.com/davisidarta/dbMAP/blob/master/dbmap/ann.py), which can be used similarly to sklearn.neighbors (it's built upon sklearn estimator class), perhaps that could be useful. As I'm on my own (first and perhaps only person in Brazil doing single-cell analysis, to the best of my knowledge), I didn't have the time yet to make a push request to scanpy of either the approximate nearest neighbor's wrapper or the diffusion basis. Perhaps we could work together to add them to scanpy?  The diffusion basis is nearly as fast as PCA, does not require scaling, uses less RAM, and automatically estimates the intrinsic data dimensionality by selecting only meaningful components (i.e. with evals > 0).

_Talking about CellRank, you could easily use that with a KNN graph computed from NMSLIB in whatever non-metric space, we take whatever is saved in adata.p as our input to compute transition matrices. Re your question of using CellRank without velocities - this is an excellent question, and it's the reason for our modular design of CellRank, structuring it into kernels and estimators. Basically, kernels build transition matrices and estimators do fun stuff with them. Our most widely used kernel is the velocity_kernel, however, we also have a palantir_kernel which mimicks palantir. For that, you don't need velocity information, just any pseudotime will do! Further kernels are likely to come (think of e.g. including multi-model information in transition matrix estimation), and you can even build your own kernel, as has recently been done here, allowing static optimal transport to be used from within CellRank!_

This is great! I'll definitely check out how results look like if using the Palantir kernel, without RNA velocity information. On the other hand, dbMAP also implements a new anisotropic, adaptive-alpha-decaying kernel which I found to be a little more generalized than Palantir to compute transition matrices, and these can be easily built by setting the 'transitions' parameter to True, so maybe I'll look into adding it as an external kernel, if that's possible.

 _Oh, and thanks for offering to share that data set! However, that's the one where you don't have velocity information, right?_

Gladly - we didn't make them public earlier because the files are huge (~30GB for the 130k human lung atlas, ~12GB for the alveolar cells). Unfortunately, we don't have velocity information for this one. However, you could map your data to this reference to get the full lineage topology, and then use only your data with RNA velocity information to project state shifts. I'm compressing the .h5ad file (it's darn big) and will upload it to https://figshare.com/articles/dataset/Integrated_Human_Lung_Cell_Atlas/12470162 asap. Hope it is useful!

**Comment by davisidarta** [comment]
Hi @Marius1311 ! I finally managed to upload it [here](https://mab.to/jiqcw4Iv2). Hope it is useful!

**Comment by Marius1311** [comment]
Thanks @davisidarta for sharing the file. As to the scanpy integration, I don't have capacity for that at the moment, sorry. 

As to the potential new kernel, that sounds interesting and that's something we could discuss in a separate issue in the [CellRank](https://cellrank.readthedocs.io/en/latest/) repo. 

If any issues come up with the palantir-kernel as it is on your data, feel free to open a new issue in the cellrank repo!
